modules/http/server/src/test/scala/Utils.scala

```diff
@@ -0,0 +1,118 @@
+  }
+
+  implicit class RequestOps[F[_]](request: Request[F]) {
+
+    def asStream[A](implicit decoder: Decoder[A]): Stream[F, A] =
+      request
+        .jsonBodyAsStream[A]
+        .adaptError { // mimic behavior of MessageOps.as[T] in handling of parsing errors
+          case ex: ParseException =>
+            MalformedMessageBodyFailure(ex.getMessage, Some(ex)) // will return 400 instead of 500
+        }
+  }
+
+  implicit class ResponseOps[F[_]](response: Response[F]) {
+
+    implicit private val throwableDecoder: Decoder[Throwable] =
+      Decoder.decodeTuple2[String, String].map {
+        case (cls, msg) =>
+          Class
+            .forName(cls)
+            .getConstructor(classOf[String])
+            .newInstance(msg)
+            .asInstanceOf[Throwable]
+      }
+
+    def asStream[A](implicit decoder: Decoder[A]): Stream[F, A] =
+      if (response.status.code != 200) Stream.raiseError(ResponseError(response.status))
+      else response.jsonBodyAsStream[Either[Throwable, A]].rethrow
+  }
+
+  implicit class Fs2StreamOps[F[_], A](stream: Stream[F, A]) {
+
+    implicit private val throwableEncoder: Encoder[Throwable] = new Encoder[Throwable] {
+      def apply(ex: Throwable): Json = (ex.getClass.getName, ex.getMessage).asJson
+    }
+
+    def asJsonEither(implicit encoder: Encoder[A]): Stream[F, Json] = stream.attempt.map(_.asJson)
+
+    def toObservable(implicit F: Effect[F], ec: ExecutionContext): Observable[A] =
+      Observable.fromReactivePublisher(stream.toUnicastPublisher)
+  }
+
+  implicit class MonixStreamOps[A](stream: Observable[A]) {
+
+    def toFs2Stream[F[_]](implicit F: Effect[F], sc: Scheduler): Stream[F, A] =
+      stream.toReactivePublisher.toStream[F]
+  }
+
+  implicit class FResponseOps[F[_]: Sync](response: F[Response[F]]) extends Http4sDsl[F] {
+
+    def adaptErrors: F[Response[F]] = response.handleErrorWith {
+      case se: StatusException         => errorFromStatus(se.getStatus, se.getMessage)
+      case sre: StatusRuntimeException => errorFromStatus(sre.getStatus, sre.getMessage)
+      case other: Throwable            => InternalServerError(other.getMessage)
+    }
+
+    private def errorFromStatus(status: io.grpc.Status, message: String): F[Response[F]] =
+      status.getCode match {
+        case INVALID_ARGUMENT  => BadRequest(message)
+        case UNAUTHENTICATED   => BadRequest(message)
+        case PERMISSION_DENIED => Forbidden(message)
+        case NOT_FOUND         => NotFound(message)
+        case UNAVAILABLE       => ServiceUnavailable(message)
+        case _                 => InternalServerError(message)
+      }
+  }
+
+  def handleResponseError[F[_]: Sync](errorResponse: Response[F]): F[Throwable] =
+    errorResponse.bodyAsText.compile.foldMonoid.map(body =>
+      ResponseError(errorResponse.status, Some(body).filter(_.nonEmpty)))
+}
+
+case class ResponseError(status: Status, msg: Option[String] = None)
```

Comment:
`final case class`

Reply:
Seems like there's two schools of thought about this, not sure if we have a standard. I'm fine with adding it.